nativescript-3.2.0-memory-management/README.md

```diff
@@ -0,0 +1,116 @@
+
+**GC** - garbage collection in general. When GCs run, they first block the threads to collect all root instances from the stack. Then resume the execution until they mark all reachable objects in a separate thread. Then block again the threads to complete the marking. And at the end start finalizing and deallocating unreachable instances. While the actual GC implementation may be much more sophisticated, all implementations in virtual machines used for UI aim at minimizing the time the main thread is blocked. Java, iOS's JavaScriptCore and Android's V8 are 3 state of the art virtual machines used in NativeScript, that have garbage collectors.
+
+**set weak/strong** - when an instance is set strong it is ensured to survive a garbage collection along with all instances reachable from it. When an instance is set weak, it can be collected if it is no longer reachable from other live instances.
+
+**splice** - let's introduce a new NativeScript term. Splice we will call a bond, made between a native instance and its representation in JavaScript by a JavaScript instance. It is possible that the splice is instantiated first in native in the cases of the iOS AppDelegate class, Android's Application, Activity and Fragment classes.
+
+Splice includes APIs to:
+ - Get the native instance given the JavaScript instance
+ - Get the JavaScript instance given the native instance
+ - Synchronize the life-cycle of the two instances
+ - Proxy method calls to and fro the JavaScript and native instances
+ - Throw when methods on the native or the JavaScript instance are called while the other is garbage collected or deallocated
+
+When is a Splice created?
+ - Native instance is returned from a constructor, method, property, block, anonymous interface, lambda etc. called from JavaScript
+ - Native instance is passed as an argument to a constructor, method, property, block, anonymous interface, lambda etc. implemented in JavaScript
+ - Native class extended in JavaScript is instantiated either in native or JavaScript
+
+## iOS
+The Objective-C runtime has no GC but runs on ref-counting instead. The retain and release of each Objective-C instance can be intercepted and code can be injected on retain/release. The Objective-C has association API that allows native objects to be assigned dynamically key-value pairs. The JavaScriptCore has API to retain/release JavaScript instances that can be used to root/unroot them (allow or deny them from being garbage collected). The Splice will refer to linking an Objective-C instance of a class with a JavaScript instance.
+
+### Splice Live-Cycle
+When a splice is made, it increases the ref-count of the Objective-C instance by 1, and if the ref-count is >1 the splice roots the JavaScript instance. From that point on:
+ - If the Objective-C instance ref-count is incremented from 1 to 2, the JavaScript instance is set strong.
```

Comment:
No, these actions happen when the ref count changes from 1 to 2 exactly or from 2 to 1. It won't happen if it is further incremented.